scripts/plugins/extauth-hook-AD.py

```diff
@@ -9,138 +9,304 @@
 
-        if is_group:
-            os.write(self.temp_fd, "account sufficient pam_succeed_if.so user ingroup %s\n" % name)
+    def _apply_to_cache(self):
+        if self._ad_enabled:
+            if self._backend == ADBackend.bd_pbis:
+                ad_pam_module = "/lib/security/pam_lsass.so"
+            else:
+                ad_pam_module = "pam_winbind.so"
+            content = self.ad_pam_format.format(ad_module=ad_pam_module)
         else:
-            os.write(self.temp_fd, "account sufficient pam_succeed_if.so user = %s\n" % name)
+            content = self.no_ad_pam
+        self._lines = content.split("\n")
 
+class DynamicPam(ADConfig):
+    def __init__(self, session, arg, ad_enabled=True):
+        super(DynamicPam, self).__init__("/etc/pam.d/hcp_users", session, arg, ad_enabled,
+                                         load_existing=False)
 
-def rewrite_etc_pamd_ssh(session, args):
-    # Rewrite the PAM SSH config using the latest info from Active Directory
-    # and the list of subjects from xapi
-    try:
-        config = PamSshConfig()
-        subjects = session.xenapi.subject.get_all()
-        admin_role = session.xenapi.role.get_by_name_label('pool-admin')[0]
+    def _apply_to_cache(self):
+        # AD is not enabled, just return as the configure file will be removed during install
+        if not self._ad_enabled:
+            return
+        # Rewrite the PAM SSH config using the latest info from Active Directory
+        # and the list of subjects from xapi
+        subjects = self._session.xenapi.subject.get_all()
+        admin_role = self._session.xenapi.role.get_by_name_label('pool-admin')[0]
         # Add each subject which contains the admin role
         for opaque_ref in subjects:
-            subject_rec = session.xenapi.subject.get_record(opaque_ref)
+            subject_rec = self._session.xenapi.subject.get_record(opaque_ref)
             sid = subject_rec['subject_identifier']
+            name = subject_rec["other_config"]["subject-name"]
+            is_group = subject_rec["other_config"]["subject-is-group"] == "true"
+            logger.debug("Permit %s with sid %s is_group as %s", name, sid, is_group)
             if admin_role in subject_rec['roles']:
-                config.add_subject(sid)
-        config.install()
+                self._add_subject(name, is_group)
+
+    def _add_subject(self, name, is_group):
+        condition = "ingroup {}".format(name) if is_group else "= {}".format(name)
+        self._lines.append("account sufficient pam_succeed_if.so user {} ".format(condition))
+
+    def _install(self):
+        if self._ad_enabled:
+            super(DynamicPam, self)._install()
+        else:
+            if os.path.exists(self._file_path):
+                os.remove(self._file_path)
+
+
+class KeyValueConfig(ADConfig):
+    _empty_line_prefix = "__key_vluae_config_empty_line_prefix_" # Presume normal config does not have such keys
```

Comment:
This seems to have a typo

```suggestion
    _empty_line_prefix = "__key_value_config_empty_line_prefix_" # Presume normal config does not have such keys
```